### PR TITLE
cmd: avoid package state in cli constructors

### PR DESCRIPTION
**Pull request by tychoish**
This began as an attempt to remove a single panic in
`libs/log/default.go` but removes a bunch of package state in the
command package.

### CODE REVIEW

File: docs/tutorials/go.md [hunk 1]
```diff
@@ -367,7 +367,11 @@ func main() {
 
  flag.Parse()
 
- logger := log.MustNewDefaultLogger(log.LogFormatPlain, log.LogLevelInfo, false)
+ logger, err := log.NewDefaultLogger(log.LogFormatPlain, log.LogLevelInfo, false)
+ if err != nil {
```
Comment (creachadair) [inline docs/tutorials/go.md:371]: Did gofmt approve this?

**Review by creachadair — APPROVED** [review]

File: docs/tutorials/go.md (hunk 1 above)
Comment (tychoish) [reply docs/tutorials/go.md:371]: it's in a markdown file

File: cmd/priv_val_server/main.go [hunk 2]
```diff
@@ -45,10 +45,15 @@ func main() {
...
 	)
+
+	logger, err := log.NewDefaultLogger(log.LogFormatPlain, log.LogLevelInfo)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "failed to construct logger: %v", err)
+		os.Exit(1)
+	}
+	logger = logger.With("module", "priv_val")
+
 	flag.Parse()
```
Comment (creachadair) [inline cmd/priv_val_server/main.go:57]: Optional, just for keeping like with like.

```suggestion
	flag.Parse()

	logger, err := log.NewDefaultLogger(log.LogFormatPlain, log.LogLevelInfo)
	if err != nil {
		fmt.Fprintf(os.Stderr, "failed to construct logger: %v", err)
		os.Exit(1)
	}
	logger = logger.With("module", "priv_val")

```

File: cmd/tendermint/commands/init.go [hunk 3]
```diff
@@ -7,51 +7,50 @@ import (
...
-	ValidArgs: []string{"full", "validator", "seed"},
-	// We allow for zero args so we can throw a more informative error
-	Args: cobra.MaximumNArgs(1),
-	RunE: initFiles,
-}
-
-var (
-	keyType string
-)
+// MakeInitFilesCommand initializes a fresh Tendermint Core instance.
```
Comment (creachadair) [inline cmd/tendermint/commands/init.go:19]: I think this comment is subtly no longer quite accurate:
```suggestion
// MakeInitFilesCommand returns the command to initialize a fresh Tendermint Core instance.
```

File: cmd/tendermint/commands/inspect.go [hunk 4]
```diff
@@ -6,14 +6,17 @@ import (
 
 	"github.com/spf13/cobra"
 
+	"github.com/tendermint/tendermint/config"
 	"github.com/tendermint/tendermint/internal/inspect"
+	"github.com/tendermint/tendermint/libs/log"
 )
 
 // InspectCmd is the command for starting an inspect server.
```
Comment (creachadair) [inline cmd/tendermint/commands/inspect.go:14]:
```suggestion
// InspectCmd constructs the command to start an inspect server.
```

File: cmd/tendermint/commands/light.go [hunk 5]
```diff
@@ -25,10 +26,57 @@ import (
 )
 
 // LightCmd represents the base command when called without any subcommands
```
Comment (creachadair) [inline cmd/tendermint/commands/light.go:28]:
```suggestion
// LightCmd constructs the base command called when invoked without any subcommands.
```

File: cmd/tendermint/commands/light.go [hunk 6]
```diff
@@ -25,10 +26,57 @@ import (
...
+		trustLevelStr  string
+
+		logLevel  string
+		logFormat string
+
+		primaryKey   = []byte("primary")
+		witnessesKey = []byte("witnesses")
+	)
+
+	checkForExistingProviders := func(db dbm.DB) (string, []string, error) {
```
Comment (creachadair) [inline cmd/tendermint/commands/light.go:51]: (optional) I think we could probably leave `primaryKey` and `witnessKey`, and a couple of these helper functions, as top-level declarations. 

The keys are morally constant, and as best I can tell do not get changed. The functions don't close over any state, and having them at the top level might help make this big function a little easier to read.

File: cmd/tendermint/commands/light.go [hunk 7]
```diff
@@ -46,185 +94,134 @@ When /abci_query is called, the Merkle key path format is:
...
 for applications built w/ Cosmos SDK).
 `,
-	RunE: runProxy,
-	Args: cobra.ExactArgs(1),
-	Example: `light cosmoshub-3 -p http://52.57.29.196:26657 -w http://public-seed-node.cosmoshub.certus.one:26657
+		RunE: func(cmd *cobra.Command, args []string) error {
+			chainID = args[0]
+			logger.Info("Creating client...", "chainID", chainID)
+
+			witnessesAddrs := []string{}
```
Comment (creachadair) [inline cmd/tendermint/commands/light.go:101]: Cosmetic nit:
```suggestion
			var witnessesAddrs []string
```

File: cmd/tendermint/commands/reindex_event.go [hunk 8]
```diff
@@ -27,58 +28,70 @@ const (
 )
 
 // ReIndexEventCmd allows re-index the event by given block height interval
```
Comment (creachadair) [inline cmd/tendermint/commands/reindex_event.go:30]:
```suggestion
// MakeReindexEventCommand constructs a command to re-index events in a block height interval.
```

File: cmd/tendermint/commands/reindex_event.go [hunk 9]
```diff
@@ -27,58 +28,70 @@ const (
...
-	},
-}
+			riArgs := eventReIndexArgs{
+				startHeight: startHeight,
+				endHeight:   endHeight,
+				sinks:       es,
+				blockStore:  bs,
+				stateStore:  ss,
+			}
+			if err = eventReIndex(cmd, riArgs); err != nil {
```
Comment (creachadair) [inline cmd/tendermint/commands/reindex_event.go:83]:
```suggestion
			if err := eventReIndex(cmd, riArgs); err != nil {
```

File: cmd/tendermint/commands/reindex_event.go [hunk 10]
```diff
@@ -223,40 +243,45 @@ func eventReIndex(cmd *cobra.Command, es []indexer.EventSink, bs state.BlockStor
 	return nil
 }
 
-func checkValidHeight(bs state.BlockStore) error {
+type checkValidHeightArgs struct {
```
Comment (creachadair) [inline cmd/tendermint/commands/reindex_event.go:246]: This one would probably be fine to pass directly (I imagine it's unlikely to change):

```go
func checkValidHeight(bs state.BlockStore, startHeight, endHeight int64) error {
```

File: cmd/tendermint/commands/replay.go [hunk 11]
```diff
@@ -2,24 +2,30 @@ package commands
 
 import (
 	"github.com/spf13/cobra"
+	"github.com/tendermint/tendermint/config"
 	"github.com/tendermint/tendermint/internal/consensus"
+	"github.com/tendermint/tendermint/libs/log"
 )
 
 // ReplayCmd allows replaying of messages from the WAL.
```
Comment (creachadair) [inline cmd/tendermint/commands/replay.go:10]:
```suggestion
// MakeReplayCommand constructs a command to replay messages from the WAL into consensus.
```

File: cmd/tendermint/commands/replay.go [hunk 12]
```diff
@@ -2,24 +2,30 @@ package commands
...
+		Use:   "replay",
+		Short: "Replay messages from WAL",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return consensus.RunReplayFile(cmd.Context(), logger, conf.BaseConfig, conf.Consensus, false)
+		},
+	}
 }
 
 // ReplayConsoleCmd allows replaying of messages from the WAL in a
 // console.
```
Comment (creachadair) [inline cmd/tendermint/commands/replay.go:22]:
```suggestion
// MakeReplayConsoleCommand constructs a command to replay WAL messages to stdout.
```

File: cmd/tendermint/commands/reset_priv_validator.go [hunk 13]
```diff
@@ -13,43 +14,49 @@ import (
 
 // ResetAllCmd removes the database of this Tendermint core
 // instance.
```
Comment (creachadair) [inline cmd/tendermint/commands/reset_priv_validator.go:16]:
```suggestion
// MakeResetAllCommand constructs a command that removes the database of
// the specified Tendermint core instance.
```

File: cmd/tendermint/commands/reset_priv_validator.go [hunk 14]
```diff
@@ -13,43 +14,49 @@ import (
...
 // XXX: this is totally unsafe.
 // it's only suitable for testnets.
-func resetPrivValidator(cmd *cobra.Command, args []string) error {
-	return resetFilePV(config.PrivValidator.KeyFile(), config.PrivValidator.StateFile(), logger)
-}
 
 // ResetAll removes address book files plus all data, and resets the privValdiator data.
 // Exported so other CLI tools can use it.
-func ResetAll(dbDir, privValKeyFile, privValStateFile string, logger log.Logger) error {
+func ResetAll(dbDir, privValKeyFile, privValStateFile string, logger log.Logger, keyType string) error {
```
Comment (creachadair) [inline cmd/tendermint/commands/reset_priv_validator.go:59]: Could maybe be unexported?

File: cmd/tendermint/commands/root.go [hunk 15]
```diff
@@ -2,65 +2,62 @@ package commands
...
-		config, err = ParseConfig()
-		if err != nil {
-			return err
-		}
-
-		logger, err = log.NewDefaultLogger(config.LogFormat, config.LogLevel)
-		if err != nil {
-			return err
-		}
+func RootCommand(conf *config.Config, logger log.Logger) *cobra.Command {
```
Comment (creachadair) [inline cmd/tendermint/commands/root.go:34]:
```suggestion
// RootCommand constructs the root command-line entry point for Tendermint core.
func RootCommand(conf *config.Config, logger log.Logger) *cobra.Command {
```

File: cmd/tendermint/commands/run_node.go [hunk 16]
```diff
@@ -20,17 +21,17 @@ var (
 
 // AddNodeFlags exposes some common configuration options on the command-line
 // These are exposed for convenience of commands embedding a tendermint node
```
Comment (creachadair) [inline cmd/tendermint/commands/run_node.go:23]:
```suggestion
// AddNodeFlags exposes some common configuration options from conf in the flag
// set for cmd. This is a convenience for commands embedding a Tendermint node.
```

File: cmd/tendermint/commands/show_node_id.go [hunk 17]
```diff
@@ -4,21 +4,22 @@ import (
...
+	"github.com/tendermint/tendermint/config"
 )
 
-// ShowNodeIDCmd dumps node's ID to the standard output.
-var ShowNodeIDCmd = &cobra.Command{
-	Use:   "show-node-id",
-	Short: "Show this node's ID",
-	RunE:  showNodeID,
-}
+// MakeShowNodeIDCmd dumps node's ID to the standard output.
```
Comment (creachadair) [inline cmd/tendermint/commands/show_node_id.go:10]:
```suggestion
// MakeShowNodeIDCommand constructs a command to dump the node ID to stdout.
```

File: cmd/tendermint/commands/show_validator.go [hunk 18]
```diff
@@ -6,75 +6,78 @@ import (
...
 	"github.com/tendermint/tendermint/crypto"
 	"github.com/tendermint/tendermint/internal/jsontypes"
+	"github.com/tendermint/tendermint/libs/log"
 	tmnet "github.com/tendermint/tendermint/libs/net"
 	tmos "github.com/tendermint/tendermint/libs/os"
 	"github.com/tendermint/tendermint/privval"
 	tmgrpc "github.com/tendermint/tendermint/privval/grpc"
 )
 
 // ShowValidatorCmd adds capabilities for showing the validator info.
```
Comment (creachadair) [inline cmd/tendermint/commands/show_validator.go:19]:
```suggestion
// MakeShowValidatorCommand constructs a command to show the validator info.
```

File: cmd/tendermint/commands/testnet.go [hunk 19]
```diff
@@ -13,76 +13,41 @@ import (
...
-		"manually override all hostnames of validators and non-validators (use --hostname multiple times for multiple hosts)")
-	TestnetFilesCmd.Flags().IntVar(&p2pPort, "p2p-port", 26656,
-		"P2P Port")
-	TestnetFilesCmd.Flags().BoolVar(&randomMonikers, "random-monikers", false,
-		"randomize the moniker for each generated node")
-	TestnetFilesCmd.Flags().StringVar(&keyType, "key", types.ABCIPubKeyTypeEd25519,
-		"Key type to generate privval file with. Options: ed25519, secp256k1")
-}
-
 // TestnetFilesCmd allows initialisation of files for a Tendermint testnet.
```
Comment (creachadair) [inline cmd/tendermint/commands/testnet.go:27]:
```suggestion
// MakeTestnetFilesCommand constructs a command to generate testnet config files.
```

File: cmd/tendermint/main.go [hunk 20]
```diff
@@ -1,37 +1,49 @@
...
-		cmd.VersionCmd,
-		cmd.InspectCmd,
-		cmd.RollbackStateCmd,
-		cmd.MakeKeyMigrateCommand(),
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	conf, err := commands.ParseConfig()
+	if err != nil {
+		panic(err)
```
Comment (creachadair) [inline cmd/tendermint/main.go:19]: I realize some of this was pre-existing, but I think these probably ought to be `stdlog.Fatal`, e.g.,

```go
stdlog.Fatalf("Reading configuration file: %v", err)
```

File: docs/tutorials/go.md (hunk 1 above)
Comment (creachadair) [reply docs/tutorials/go.md:371]: Ha. I often use play.golang.org to format samples for MD, FWIW.

**Review by creachadair — APPROVED** [review]
A few optional suggestions, otherwise LGTM. This turned out to be tricker than it appeared initially.

File: cmd/tendermint/main.go (hunk 20 above)
Comment (tychoish) [reply cmd/tendermint/main.go:19]: I had the same thought, and opted to avoid doing it in the short term, but definitely open to it.

File: cmd/tendermint/commands/reindex_event.go (hunk 10 above)
Comment (tychoish) [reply cmd/tendermint/commands/reindex_event.go:246]: yeah, I think either is fine, but I have a minor bias against having sequences of arguments of the same type that can easily be transposed.

File: cmd/tendermint/commands/reset_priv_validator.go (hunk 14 above)
Comment (tychoish) [reply cmd/tendermint/commands/reset_priv_validator.go]: yep!

File: cmd/tendermint/commands/light.go [hunk 21]
```diff
@@ -46,185 +94,134 @@ When /abci_query is called, the Merkle key path format is:
...
+				return fmt.Errorf("can't parse trust level: %w", err)
+			}
+
+			options := []light.Option{light.Logger(logger)}
+
+			if sequential {
+				options = append(options, light.SequentialVerification())
+			} else {
+				options = append(options, light.SkippingVerification(trustLevel))
+			}
```
Comment (williambanfield) [inline cmd/tendermint/commands/light.go:141]: Not from you I don't believe. 
```suggestion

			vo := light.SkippingVerification(trustLevel)
			if sequential {
				vo = light.SequentialVerification()
			}
			options = append(options, vo)

```

File: cmd/tendermint/commands/reindex_event.go [hunk 22]
```diff
@@ -27,58 +28,70 @@ const (
...
+				return
+			}
 
-		if err = eventReIndex(cmd, es, bs, ss); err != nil {
-			fmt.Println(reindexFailed, err)
-			return
-		}
+			es, err := loadEventSinks(conf)
+			if err != nil {
+				fmt.Println(reindexFailed, err)
```
Comment (williambanfield) [inline cmd/tendermint/commands/reindex_event.go:72]: Not you, but, any reason not to use `RunE` and return error here instead? `Println` seems a little odd. Might as well correct while we're here.

File: cmd/tendermint/commands/reindex_event.go [hunk 23]
```diff
@@ -27,58 +28,70 @@ const (
...
+			if err = eventReIndex(cmd, riArgs); err != nil {
+				fmt.Println(reindexFailed, err)
+				return
+			}
 
-var (
-	startHeight int64
-	endHeight   int64
-)
+			fmt.Println("event re-index finished")
```
Comment (williambanfield) [inline cmd/tendermint/commands/reindex_event.go:88]:
```suggestion
			logger.Info("event re-index finished")
```

File: cmd/tendermint/commands/reindex_event.go [hunk 24]
```diff
@@ -192,19 +212,19 @@ func eventReIndex(cmd *cobra.Command, es []indexer.EventSink, bs state.BlockStor
 			if e.NumTxs > 0 {
 				batch = indexer.NewBatch(e.NumTxs)
 
-				for i, tx := range b.Data.Txs {
+				for i := range b.Data.Txs {
 					tr := abcitypes.TxResult{
 						Height: b.Height,
 						Index:  uint32(i),
-						Tx:     tx,
+						Tx:     b.Data.Txs[i],
```
Comment (williambanfield) [inline cmd/tendermint/commands/reindex_event.go:219]: Why this change?

File: cmd/tendermint/main.go (hunk 20 above)
Comment (creachadair) [reply cmd/tendermint/main.go:19]: Definitely orthogonal to your immediate purpose, and not urgent.

File: cmd/tendermint/commands/testnet.go [hunk 25]
```diff
@@ -93,205 +58,274 @@ Example:
...
-			return err
-		}
+			fmt.Printf("Successfully initialized %v node directories\n", nValidators+nNonValidators)
+			return nil
+		},
 	}
 
-	fmt.Printf("Successfully initialized %v node directories\n", nValidators+nNonValidators)
-	return nil
+	cmd.Flags().IntVar(&nValidators, "v", 4,
```
Comment (williambanfield) [inline cmd/tendermint/commands/testnet.go:228]: I might change this to declare the `cmd` struct first, then the vars that will be used by the command function, then add the flags to the cmd struct and finally, add the command function to the cmd struct.

File: cmd/tendermint/commands/testnet.go [hunk 26]
```diff
@@ -93,205 +58,274 @@ Example:
...
+	startingIPAddr   string
+	hostnamePrefix   string
+	hostnameSuffix   string
+	randomMonikers   bool
 }
 
-func hostnameOrIP(i int) string {
-	if len(hostnames) > 0 && i < len(hostnames) {
-		return hostnames[i]
+func (args *testnetPeerArgs) hostnameOrIP(i int) (string, error) {
```
Comment (williambanfield) [inline cmd/tendermint/commands/testnet.go:280]: This only needs 3 of the args, I think keeping it a function is fine. 
```suggestion
func hostnameOrIP(i int, hostnames []string, startingIpAddr string) (string, error) {
```

File: cmd/tendermint/commands/testnet.go [hunk 27]
```diff
@@ -93,205 +58,274 @@ Example:
...
 }
 
 // get an array of persistent peers
-func persistentPeersArray(config *cfg.Config) ([]string, error) {
-	peers := make([]string, nValidators+nNonValidators)
-	for i := 0; i < nValidators+nNonValidators; i++ {
-		nodeDir := filepath.Join(outputDir, fmt.Sprintf("%s%d", nodeDirPrefix, i))
+func persistentPeersArray(config *cfg.Config, args testnetPeerArgs) ([]string, error) {
+	peers := make([]string, args.numValidators+args.numNonValidators)
+	for i := 0; i < args.numValidators+args.numNonValidators; i++ {
```
Comment (williambanfield) [inline cmd/tendermint/commands/testnet.go:303]:
```suggestion
	for i := 0; i < len(peers); i++ {
```

File: cmd/tendermint/commands/reindex_event.go (hunk 24 above)
Comment (tychoish) [reply cmd/tendermint/commands/reindex_event.go:219]: to avoid hanging on to a stale reference

File: cmd/tendermint/commands/testnet.go (hunk 26 above)
Comment (tychoish) [reply cmd/tendermint/commands/testnet.go:280]: I unwound this change because it used a couple more args afterall.

**Comment by williambanfield** [comment]
Still not totally sure I understand why we're using `fmt.Println` for errors instead of returning and error through a `RunE`?
